### PR TITLE
Namespace quirks

### PR DESCRIPTION
**Pull request by fschlimb**
continuation of #1115 (rebasing my branch seemed to confuse CI)

### CODE REVIEW

**Comment by fschlimb** [comment]
This is the continuation of #1115; I opened a new PR because rebasing my branch seemed to confuse CI.

I simplified the code and also reduced to changes related to the template/namespace issue. Most of the removed pieces from #1115 went into #1488.

I added a more extensive test-case also covering the case from #810.  I will check with our project to see if the new SWIG code-base together with this makes #810 unnecessary. I'll leave #810 open until I clarified. 

@wsfulton @ojwb Any other comments/thoughts/requests?

File: Source/CParse/templ.c [hunk 1]
```diff
@@ -371,11 +370,20 @@ int Swig_cparse_template_expand(Node *n, String *rname, ParmList *tparms, Symtab
...
-	    SwigType_typename_replace(s, name, dvalue);
-	    SwigType_typename_replace(s, tbase, iname);
-	    /*      Printf(stdout,"'%s'\n", s); */
+	    /*
+	      The approach of 'trivially' replacing template arguments is kind of fragile.
+	      In particular if types with similar name in different namespaces appear.
+	      We will not replace template args if a type/class exists with the same
+	      name which is not a template.
+	    */
+	    Node * __n = Swig_symbol_clookup(s, 0);
```
Comment (ojwb) [inline Source/CParse/templ.c]: Identifiers containing a double underscore anywhere are reserved (https://en.cppreference.com/w/cpp/language/identifiers) so these variables really need renaming to not use reserved names.

File: Source/CParse/templ.c [hunk 2]
```diff
@@ -371,11 +370,20 @@ int Swig_cparse_template_expand(Node *n, String *rname, ParmList *tparms, Symtab
...
+	    String * __nn = __n ? Getattr(__n, "sym:name") : 0;
+	    String * nn = __nn ? Getattr(n, "sym:name") : 0;
+	    int __eq = __nn && nn ? Strcmp(nn, __nn) : 1;
```
Comment (ojwb) [inline Source/CParse/templ.c]: This is confusing code.  The variable is called `__eq` which suggests it's true if equal, but (when both the arguments are non-NULL) it's set to the result of `Strcmp()`, which is `0` if equal and non-zero if non-equal.  So if the strings are equal, it seems `__eq` is false!?!

Better to compare the strings with `Equal(a, b)` and make the sense of the variable match what its name suggests.

File: Source/CParse/templ.c [hunk 3]
```diff
@@ -371,11 +370,20 @@ int Swig_cparse_template_expand(Node *n, String *rname, ParmList *tparms, Symtab
...
+	    if(__eq!=0 || Getattr(__n, "templatetype")) {
```
Comment (ojwb) [inline Source/CParse/templ.c]: The whitespace here doesn't match the SWIG code (should have a space after `if` and spaces around `!=`).

File: Source/CParse/templ.c (hunk 1 above)
Comment (fschlimb) [reply Source/CParse/templ.c]: done

File: Source/CParse/templ.c (hunk 2 above)
Comment (fschlimb) [reply Source/CParse/templ.c]: done

File: Source/CParse/templ.c (hunk 3 above)
Comment (fschlimb) [reply Source/CParse/templ.c]: done

**Comment by fschlimb** [comment]
@ojwb thanks for your review. I made the suggested changes.

**Comment by ojwb** [comment]
Thanks.  Looks OK to me now, but I think it would be good for @wsfulton to review too.

**Comment by fschlimb** [comment]
@wsfulton, any chane this will ever make it to master?

**Comment by ojwb** [comment]
I've pushed a minor whitespace tweak to the branch (the other pre-function comment blocks have a line after them before the function, at least in this file, so removing this one seems undesirable) which should also mean we'll get up to date CI results, but annoyingly a whole slew of builds seem to have failed for some sort of unrelated CI system reason.  

I'm hesitant to merge this without a thumbs up from @wsfulton as I'm not quite sure I've got my head around exactly what the problem is that this addresses.  From this comment I gather it's to do with doing a textual search and replace to expand template arguments but I haven't fully understood what goes wrong:

```
              The approach of 'trivially' replacing template arguments is kind of fragile.
	      In particular if types with similar name in different namespaces appear.
	      We will not replace template args if a type/class exists with the same
	      name which is not a template.
```

Maybe we can make the substitution a bit smarter for a better fix.

**Comment by fschlimb** [comment]
I guess the fundamental problem is that SWIG tries to implement full C++ compiler frontend and uses such "trivial" replacements at various places (not only here). I am actually surprised how well this works for so many codes. Chapeau! Template and template argument substitution is a very complex issue with many subtleties. Trying to solve this would probably require a fundamentally different approach.
As far as I remember this fix does not introduce any regression (in the test suite) but makes SWIG support a larger set of possible namespace/template constructs.

**Comment by wsfulton** [comment]
I've added this to the swig-4.1 milestone for review before we release the next version.

**Comment by wsfulton** [comment]
fa00622614d35716b429b12aa9dc1bde8e4b586d contains the majority of this patch, that is, the changes in templ.c and the test case. The only observable fix I could see from the testcase was for Bar:
```c++
              namespace A { class Bar {}; }
              namespace B { template<typename T, typename U> class Bar : public A::Bar {}; }
 ```
I couldn't see any observable difference for the changes in typepass.cxx in either the test-suite nor the additional test case here; the change doesn't look unreasonable, but I'd like an example of the specialization that requires the change in order to accept it.

**Comment by fschlimb** [comment]
Are you saying that the test here works correctly with fa00622?

**Comment by wsfulton** [comment]
Yes, fa00622 includes the testcase from this pull request.

**Pull request closed without merging**